ouroboros-consensus/src/Ouroboros/Consensus/Fragment/Validated.hs

```diff
@@ -0,0 +1,65 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+-- | Intended for qualified import
+--
+-- > import Ouroboros.Consensus.Fragment.Validated (ValidatedFragment)
+-- > import qualified Ouroboros.Consensus.Fragment.Validated as VF
+module Ouroboros.Consensus.Fragment.Validated (
+    ValidatedFragment -- Opaque
+  , validatedFragment
+  , validatedLedger
+  , new
+  ) where
+
+import           GHC.Stack
+
+import           Ouroboros.Network.AnchoredFragment (AnchoredFragment)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           Ouroboros.Network.Block
+
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Util.Assert
+
+-- | Validated chain fragment along with the ledger state after validation
+--
+-- INVARIANT:
+--
+-- > AF.headPoint validatedFragment == ledgerTipPoint validatedLedger
+data ValidatedFragment l blk = ValidatedFragment {
+    -- | Chain fragment
+    validatedFragment :: !(AnchoredFragment (Header blk))
+
+    -- | Ledger after after validation
+  , validatedLedger   :: !l
+  }
+
+invariant :: forall l blk. ApplyBlock l blk
+          => ValidatedFragment l blk -> Either String ()
+invariant ValidatedFragment{..}
+    | ledgerTip /= headPoint
+    = Left $ show ledgerTip ++ " /= " ++ show headPoint
```

Comment:
Would it be useful to include which is which in the error message?

Reply:
Will do.